base/inference.jl

```diff
@@ -1479,7 +1529,20 @@ function typeinf(linfo::LambdaStaticData,atypes::Tuple,sparams::Tuple, def, cop)
                 end
             end
             pc´ = pc+1
-            if isa(stmt,GotoNode)
+            if isa(changes,StateUpdate) && isa((changes::StateUpdate).var, GenSym)
+                changes = changes::StateUpdate
+                id = (changes.var::GenSym).id+1
+                new = changes.vtype
+                old = gensym_types[id]
+                if old===NF || !(new <: old)
+                    gensym_types[id] = tmerge(old, new)
+                    for r in gensym_uses[id]
+                        if !is(s[r],()) # s[r] === () => unreached statement
```

Comment:
part of the point of the GenSym implementation is that their information is flow-insensitive. they are only auto-generated in places where we know, a-priori, that this optimization is valid. they aren't really accessible to be created by the user (since they don't have a syntax). i could lazy-populate the `gensym_uses` vector (i should switch it to an IntSet), but I'm not sure of that value in that